server/src/main/java/org/elasticsearch/common/metrics/Counters.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.metrics;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Simple usage stat counters based on longs. Internally this is a map mapping from String to a CounterMetric.
+ * Calling toNestedMap() will create a nested map, where each dot of the key name will nest deeper. This class allows the
+ * stats producer to not worry about how the map is actually nested.
+ */
+public class Counters implements Writeable {
+
+    private final Map<String, CounterMetric> counters = new ConcurrentHashMap<>();
```

Comment:
Hm, not sure, I do think that the thread safety is a requirement of this class, for this reason I would be open in using it also in the definition. I had my doubts because any thread safe map could be used here theoretically. What are your thoughts on the topic?

Reply:
>I had my doubts because any thread safe map could be used here theoretically. What are your thoughts on the topic?

This would be expressed by the `ConcurrentMap` interface right? Do you think it's too restrictive? (ie. is there a possible implementation that implements `Map` but not `ConcurrentMap` that we might use here)

Reply:
Ah sorry... I read `ConcurrentHashMap`.... Definitely, I will change it.